TestUpdateKibanaSecureSettings is flaky

https://devops-ci.elastic.co/job/cloud-on-k8s-e2e-tests-snapshot-versions/106/testReport/

```
=== RUN   TestUpdateKibanaSecureSettings/secure_settings_should_eventually_be_set_in_all_nodes_keystore
Retries (5m0s timeout): ................................................................................
    utils.go:84: 
        	Error Trace:	utils.go:84
        	Error:      	Received unexpected error:
        	            	invalid keystore entries. Expected: [logging.verbose]. Actual: []
        	Test:       	TestUpdateKibanaSecureSettings/secure_settings_should_eventually_be_set_in_all_nodes_keystore
```

I can reproduce the issue locally with the latest version of `7.9.0-SNAPSHOT` :

```
=== RUN   TestUpdateKibanaSecureSettings/secure_settings_should_eventually_be_set_in_all_nodes_keystore
Retries (5m0s timeout): ...........................................................................
{"log.level":"error","@timestamp":"2020-07-28T11:03:25.392+0200","message":"stopping early","service.version":"0.0.0-00000000","service.type":"eck","ecs.version":"1.4.0","error":"test failure","error.stack_trace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/michael/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\ngithub.com/elastic/cloud-on-k8s/test/e2e/test.StepList.RunSequential\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/test/step.go:43\ngithub.com/elastic/cloud-on-k8s/test/e2e/kb.TestUpdateKibanaSecureSettings\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/test/e2e/kb/keystore_test.go:114\ntesting.tRunner\n\t/usr/local/Cellar/go/1.13.7/libexec/src/testing/testing.go:909"}
    --- FAIL: TestUpdateKibanaSecureSettings/secure_settings_should_eventually_be_set_in_all_nodes_keystore (300.00s)
        utils.go:84: 
            	Error Trace:	utils.go:84
            	Error:      	Received unexpected error:
            	            	invalid keystore entries. Expected: [logging.verbose]. Actual: []
            	Test:       	TestUpdateKibanaSecureSettings/secure_settings_should_eventually_be_set_in_all_nodes_keystore
```

I have the following event emitted on the cluster: 

```
default       0s          Warning   OOMKilling                node/gke-michael-dev-2-default-pool-c3327033-hjlg                                Memory cgroup out of memory: Kill process 31303 (node) score 1965 or sacrifice child
Killed process 31303 (node) total-vm:2188540kB, anon-rss:1006452kB, file-rss:28584kB, shmem-rss:0kB
```

Not sure if it is related to #3344, still investigating...

--- a/pkg/controller/common/volume/shared_volumes.go
+++ b/pkg/controller/common/volume/shared_volumes.go
@@ -6,7 +6,7 @@ package volume
 
 import corev1 "k8s.io/api/core/v1"
 
-// SharedVolume between the init container and the ES container.
+// SharedVolume between the init container and the main container.
 type SharedVolume struct {
 	VolumeName             string // Volume name
 	InitContainerMountPath string // Mount path in the init container

--- a/pkg/controller/common/volume/shared_volumes.go
+++ b/pkg/controller/common/volume/shared_volumes.go
@@ -2,34 +2,38 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
-package initcontainer
+package volume
 
 import corev1 "k8s.io/api/core/v1"
 
 // SharedVolume between the init container and the ES container.
 type SharedVolume struct {
-	Name                   string // Volume name
+	VolumeName             string // Volume name
 	InitContainerMountPath string // Mount path in the init container
-	EsContainerMountPath   string // Mount path in the Elasticsearch container
+	ContainerMountPath     string // Mount path in the main container (e.g. Elasticsearch or Kibana)
 }
 
 func (v SharedVolume) InitContainerVolumeMount() corev1.VolumeMount {
 	return corev1.VolumeMount{
 		MountPath: v.InitContainerMountPath,
-		Name:      v.Name,
+		Name:      v.VolumeName,
 	}
 }
 
-func (v SharedVolume) EsContainerVolumeMount() corev1.VolumeMount {
+func (v SharedVolume) Name() string {
+	return v.VolumeName
+}
+
+func (v SharedVolume) VolumeMount() corev1.VolumeMount {
 	return corev1.VolumeMount{
-		MountPath: v.EsContainerMountPath,
-		Name:      v.Name,
+		MountPath: v.ContainerMountPath,
+		Name:      v.VolumeName,
 	}
 }
 
 func (v SharedVolume) Volume() corev1.Volume {
 	return corev1.Volume{
-		Name: v.Name,
+		Name: v.VolumeName,
 		VolumeSource: corev1.VolumeSource{
 			EmptyDir: &corev1.EmptyDirVolumeSource{},
 		},
@@ -49,10 +53,10 @@ func (v SharedVolumeArray) InitContainerVolumeMounts() []corev1.VolumeMount {
 	return mounts
 }
 
-func (v SharedVolumeArray) EsContainerVolumeMounts() []corev1.VolumeMount {
+func (v SharedVolumeArray) ContainerVolumeMounts() []corev1.VolumeMount {
 	mounts := make([]corev1.VolumeMount, len(v.Array))
 	for i, v := range v.Array {
-		mounts[i] = v.EsContainerVolumeMount()
+		mounts[i] = v.VolumeMount()
 	}
 	return mounts
 }
@@ -61,7 +65,7 @@ func (v SharedVolumeArray) Volumes() []corev1.Volume {
 	volumes := make([]corev1.Volume, len(v.Array))
 	for i, v := range v.Array {
 		volumes[i] = corev1.Volume{
-			Name: v.Name,
+			Name: v.VolumeName,
 			VolumeSource: corev1.VolumeSource{
 				EmptyDir: &corev1.EmptyDirVolumeSource{},
 			},

--- a/pkg/controller/elasticsearch/initcontainer/prepare_fs.go
+++ b/pkg/controller/elasticsearch/initcontainer/prepare_fs.go
@@ -26,28 +26,28 @@ const (
 // Volumes that are shared between the prepare-fs init container and the ES container
 var (
 	// EsBinSharedVolume contains the ES bin/ directory
-	EsBinSharedVolume = SharedVolume{
-		Name:                   "elastic-internal-elasticsearch-bin-local",
+	EsBinSharedVolume = volume.SharedVolume{
+		VolumeName:             "elastic-internal-elasticsearch-bin-local",
 		InitContainerMountPath: "/mnt/elastic-internal/elasticsearch-bin-local",
-		EsContainerMountPath:   "/usr/share/elasticsearch/bin",
+		ContainerMountPath:     "/usr/share/elasticsearch/bin",
 	}
 
 	// EsConfigSharedVolume contains the ES config/ directory
-	EsConfigSharedVolume = SharedVolume{
-		Name:                   "elastic-internal-elasticsearch-config-local",
+	EsConfigSharedVolume = volume.SharedVolume{
+		VolumeName:             "elastic-internal-elasticsearch-config-local",
 		InitContainerMountPath: "/mnt/elastic-internal/elasticsearch-config-local",
-		EsContainerMountPath:   esvolume.ConfigVolumeMountPath,
+		ContainerMountPath:     esvolume.ConfigVolumeMountPath,
 	}
 
 	// EsPluginsSharedVolume contains the ES plugins/ directory
-	EsPluginsSharedVolume = SharedVolume{
-		Name:                   "elastic-internal-elasticsearch-plugins-local",
+	EsPluginsSharedVolume = volume.SharedVolume{
+		VolumeName:             "elastic-internal-elasticsearch-plugins-local",
 		InitContainerMountPath: "/mnt/elastic-internal/elasticsearch-plugins-local",
-		EsContainerMountPath:   "/usr/share/elasticsearch/plugins",
+		ContainerMountPath:     "/usr/share/elasticsearch/plugins",
 	}
 
-	PluginVolumes = SharedVolumeArray{
-		Array: []SharedVolume{
+	PluginVolumes = volume.SharedVolumeArray{
+		Array: []volume.SharedVolume{
 			EsConfigSharedVolume,
 			EsPluginsSharedVolume,
 			EsBinSharedVolume,
@@ -59,23 +59,23 @@ var (
 		Array: []LinkedFile{
 			{
 				Source: stringsutil.Concat(esvolume.XPackFileRealmVolumeMountPath, "/", filerealm.UsersFile),
-				Target: stringsutil.Concat(EsConfigSharedVolume.EsContainerMountPath, "/", filerealm.UsersFile),
+				Target: stringsutil.Concat(EsConfigSharedVolume.ContainerMountPath, "/", filerealm.UsersFile),
 			},
 			{
 				Source: stringsutil.Concat(esvolume.XPackFileRealmVolumeMountPath, "/", user.RolesFile),
-				Target: stringsutil.Concat(EsConfigSharedVolume.EsContainerMountPath, "/", user.RolesFile),
+				Target: stringsutil.Concat(EsConfigSharedVolume.ContainerMountPath, "/", user.RolesFile),
 			},
 			{
 				Source: stringsutil.Concat(esvolume.XPackFileRealmVolumeMountPath, "/", filerealm.UsersRolesFile),
-				Target: stringsutil.Concat(EsConfigSharedVolume.EsContainerMountPath, "/", filerealm.UsersRolesFile),
+				Target: stringsutil.Concat(EsConfigSharedVolume.ContainerMountPath, "/", filerealm.UsersRolesFile),
 			},
 			{
 				Source: stringsutil.Concat(settings.ConfigVolumeMountPath, "/", settings.ConfigFileName),
-				Target: stringsutil.Concat(EsConfigSharedVolume.EsContainerMountPath, "/", settings.ConfigFileName),
+				Target: stringsutil.Concat(EsConfigSharedVolume.ContainerMountPath, "/", settings.ConfigFileName),
 			},
 			{
 				Source: stringsutil.Concat(esvolume.UnicastHostsVolumeMountPath, "/", esvolume.UnicastHostsFile),
-				Target: stringsutil.Concat(EsConfigSharedVolume.EsContainerMountPath, "/", esvolume.UnicastHostsFile),
+				Target: stringsutil.Concat(EsConfigSharedVolume.ContainerMountPath, "/", esvolume.UnicastHostsFile),
 			},
 		},
 	}

--- a/pkg/controller/elasticsearch/initcontainer/prepare_fs_script.go
+++ b/pkg/controller/elasticsearch/initcontainer/prepare_fs_script.go
@@ -8,12 +8,14 @@ import (
 	"bytes"
 	"fmt"
 	"html/template"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
 )
 
 // TemplateParams are the parameters manipulated in the scriptTemplate
 type TemplateParams struct {
 	// SharedVolumes are directories to persist in shared volumes
-	PluginVolumes SharedVolumeArray
+	PluginVolumes volume.SharedVolumeArray
 	// LinkedFiles are files to link individually
 	LinkedFiles LinkedFilesArray
 	// ChownToElasticsearch are paths that need to be chowned to the Elasticsearch user/group.
@@ -100,11 +102,11 @@ var scriptTemplate = template.Must(template.New("").Parse(
 	# so installed plugins files can to be used by the ES container
 	mv_start=$(date +%s)
 	{{range .PluginVolumes.Array}}
-		if [[ -z "$(ls -A {{.EsContainerMountPath}})" ]]; then
-			echo "Empty dir {{.EsContainerMountPath}}"
+		if [[ -z "$(ls -A {{.ContainerMountPath}})" ]]; then
+			echo "Empty dir {{.ContainerMountPath}}"
 		else
-			echo "Copying {{.EsContainerMountPath}}/* to {{.InitContainerMountPath}}/"
-			cp -av {{.EsContainerMountPath}}/* {{.InitContainerMountPath}}/
+			echo "Copying {{.ContainerMountPath}}/* to {{.InitContainerMountPath}}/"
+			cp -av {{.ContainerMountPath}}/* {{.InitContainerMountPath}}/
 		fi
 	{{end}}
 	echo "Files copy duration: $(duration $mv_start) sec."

--- a/pkg/controller/elasticsearch/nodespec/volumes.go
+++ b/pkg/controller/elasticsearch/nodespec/volumes.go
@@ -85,7 +85,7 @@ func buildVolumes(esName string, nodeSpec esv1.NodeSet, keystoreResources *keyst
 	}
 
 	volumeMounts := append(
-		initcontainer.PluginVolumes.EsContainerVolumeMounts(),
+		initcontainer.PluginVolumes.ContainerVolumeMounts(),
 		esvolume.DefaultDataVolumeMount,
 		esvolume.DefaultLogsVolumeMount,
 		usersSecretVolume.VolumeMount(),

--- a/pkg/controller/kibana/config_reconcile.go
+++ b/pkg/controller/kibana/config_reconcile.go
@@ -27,6 +27,7 @@ const (
 	ConfigVolumeMountPath              = "/usr/share/kibana/config"
 	InitContainerConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config-local"
 
+	// InternalConfigVolumeName is a volume which contains the generated configuration.
 	InternalConfigVolumeName      = "elastic-internal-kibana-config"
 	InternalConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config"
 
@@ -35,7 +36,10 @@ const (
 
 var (
 	// ConfigSharedVolume contains the Kibana config/ directory, it's an empty volume where the required configuration
-	// is initialized by the elastic-internal-init-config init container.
+	// is initialized by the elastic-internal-init-config init container. Its content is then shared by the init container
+	// that creates the keystore and the main Kibana container.
+	// This is needed in order to have in a same directory both the generated configuration and the keystore file  which
+	// is created in /usr/share/kibana/config since Kibana 7.9
 	ConfigSharedVolume = volume.SharedVolume{
 		VolumeName:             ConfigVolumeName,
 		InitContainerMountPath: InitContainerConfigVolumeMountPath,

--- a/pkg/controller/kibana/config_reconcile.go
+++ b/pkg/controller/kibana/config_reconcile.go
@@ -27,7 +27,7 @@ const (
 	ConfigVolumeMountPath              = "/usr/share/kibana/config"
 	InitContainerConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config-local"
 
-	InternalConfigVolumeName      = "elastic-internal-elasticsearch-config"
+	InternalConfigVolumeName      = "elastic-internal-kibana-config"
 	InternalConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config"
 
 	telemetryFilename = "telemetry.yml"

--- a/pkg/controller/kibana/config_reconcile.go
+++ b/pkg/controller/kibana/config_reconcile.go
@@ -23,9 +23,24 @@ import (
 
 // Constants to use for the config files in a Kibana pod.
 const (
-	VolumeName        = "config"
-	VolumeMountPath   = "/usr/share/kibana/" + VolumeName
+	ConfigVolumeName                   = "elastic-internal-kibana-config-local"
+	ConfigVolumeMountPath              = "/usr/share/kibana/config"
+	InitContainerConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config-local"
+
+	InternalConfigVolumeName      = "elastic-internal-elasticsearch-config"
+	InternalConfigVolumeMountPath = "/mnt/elastic-internal/kibana-config"
+
 	telemetryFilename = "telemetry.yml"
+)
+
+var (
+	// ConfigSharedVolume contains the Kibana config/ directory, it's an empty volume where the required configuration
+	// is initialized by the elastic-internal-init-config init container.
+	ConfigSharedVolume = volume.SharedVolume{
+		VolumeName:             ConfigVolumeName,
+		InitContainerMountPath: InitContainerConfigVolumeMountPath,
+		ContainerMountPath:     ConfigVolumeMountPath,
+	}
 )
 
 // ECK is a helper struct to marshal telemetry information.
@@ -33,18 +48,18 @@ type ECK struct {
 	ECK about.OperatorInfo `json:"eck"`
 }
 
-// SecretVolume returns a SecretVolume to hold the Kibana config of the given Kibana resource.
-func SecretVolume(kb kbv1.Kibana) volume.SecretVolume {
+// ConfigVolume returns a SecretVolume to hold the Kibana config of the given Kibana resource.
+func ConfigVolume(kb kbv1.Kibana) volume.SecretVolume {
 	return volume.NewSecretVolumeWithMountPath(
 		SecretName(kb),
-		VolumeName,
-		VolumeMountPath,
+		InternalConfigVolumeName,
+		InternalConfigVolumeMountPath,
 	)
 }
 
 // SecretName is the name of the secret that holds the Kibana config for the given Kibana resource.
 func SecretName(kb kbv1.Kibana) string {
-	return kb.Name + "-kb-" + VolumeName
+	return kb.Name + "-kb-config"
 }
 
 // ReconcileConfigSecret reconciles the expected Kibana config secret for the given Kibana resource.

--- a/pkg/controller/kibana/driver.go
+++ b/pkg/controller/kibana/driver.go
@@ -13,7 +13,6 @@ import (
 	"go.elastic.co/apm"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -36,24 +35,6 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
-
-// initContainersParameters is used to generate the init container that will load the secure settings into a keystore
-var initContainersParameters = keystore.InitContainerParameters{
-	KeystoreCreateCommand:         "/usr/share/kibana/bin/kibana-keystore create",
-	KeystoreAddCommand:            `/usr/share/kibana/bin/kibana-keystore add "$key" --stdin < "$filename"`,
-	SecureSettingsVolumeMountPath: keystore.SecureSettingsVolumeMountPath,
-	KeystoreVolumePath:            DataVolumeMountPath,
-	Resources: corev1.ResourceRequirements{
-		Requests: map[corev1.ResourceName]resource.Quantity{
-			corev1.ResourceMemory: resource.MustParse("128Mi"),
-			corev1.ResourceCPU:    resource.MustParse("100m"),
-		},
-		Limits: map[corev1.ResourceName]resource.Quantity{
-			corev1.ResourceMemory: resource.MustParse("128Mi"),
-			corev1.ResourceCPU:    resource.MustParse("100m"),
-		},
-	},
-}
 
 // minSupportedVersion is the minimum version of Kibana supported by ECK. Currently this is set to version 6.8.0.
 var minSupportedVersion = version.From(6, 8, 0)
@@ -195,6 +176,10 @@ func (d *driver) getStrategyType(kb *kbv1.Kibana) (appsv1.DeploymentStrategyType
 }
 
 func (d *driver) deploymentParams(kb *kbv1.Kibana) (deployment.Params, error) {
+	initContainersParameters, err := newInitContainersParameters(kb)
+	if err != nil {
+		return deployment.Params{}, err
+	}
 	// setup a keystore with secure settings in an init container, if specified by the user
 	keystoreResources, err := keystore.NewResources(
 		d,
@@ -207,7 +192,7 @@ func (d *driver) deploymentParams(kb *kbv1.Kibana) (deployment.Params, error) {
 		return deployment.Params{}, err
 	}
 
-	kibanaPodSpec := NewPodTemplateSpec(*kb, keystoreResources)
+	kibanaPodSpec := NewPodTemplateSpec(*kb, keystoreResources, d.buildVolumes(kb))
 
 	// Build a checksum of the configuration, which we can use to cause the Deployment to roll Kibana
 	// instances in case of any change in the CA file, secure settings or credentials contents.
@@ -220,17 +205,6 @@ func (d *driver) deploymentParams(kb *kbv1.Kibana) (deployment.Params, error) {
 	// we need to deref the secret here to include it in the checksum otherwise Kibana will not be rolled on contents changes
 	if err := commonassociation.WriteAssocsToConfigHash(d.client, kb.GetAssociations(), configChecksum); err != nil {
 		return deployment.Params{}, err
-	}
-
-	volumes := []commonvolume.SecretVolume{SecretVolume(*kb)}
-
-	if kb.AssociationConf().CAIsConfigured() {
-		esCertsVolume := esCaCertSecretVolume(*kb)
-		volumes = append(volumes, esCertsVolume)
-		for i := range kibanaPodSpec.Spec.InitContainers {
-			kibanaPodSpec.Spec.InitContainers[i].VolumeMounts = append(kibanaPodSpec.Spec.InitContainers[i].VolumeMounts,
-				esCertsVolume.VolumeMount())
-		}
 	}
 
 	if kb.Spec.HTTP.TLS.Enabled() {
@@ -246,16 +220,6 @@ func (d *driver) deploymentParams(kb *kbv1.Kibana) (deployment.Params, error) {
 		if httpCert, ok := httpCerts.Data[certificates.CertFileName]; ok {
 			_, _ = configChecksum.Write(httpCert)
 		}
-
-		httpCertsVolume := certificates.HTTPCertSecretVolume(Namer, kb.Name)
-		volumes = append(volumes, httpCertsVolume)
-	}
-
-	// attach volumes
-	kibanaContainer := GetKibanaContainer(kibanaPodSpec.Spec)
-	for _, volume := range volumes {
-		kibanaPodSpec.Spec.Volumes = append(kibanaPodSpec.Spec.Volumes, volume.Volume())
-		kibanaContainer.VolumeMounts = append(kibanaContainer.VolumeMounts, volume.VolumeMount())
 	}
 
 	// get config secret to add its content to the config checksum
@@ -285,6 +249,21 @@ func (d *driver) deploymentParams(kb *kbv1.Kibana) (deployment.Params, error) {
 		PodTemplateSpec: kibanaPodSpec,
 		Strategy:        strategyType,
 	}, nil
+}
+
+func (d *driver) buildVolumes(kb *kbv1.Kibana) []commonvolume.VolumeLike {
+	volumes := []commonvolume.VolumeLike{DataVolume, ConfigSharedVolume, ConfigVolume(*kb)}
+
+	if kb.AssociationConf().CAIsConfigured() {
+		esCertsVolume := esCaCertSecretVolume(*kb)
+		volumes = append(volumes, esCertsVolume)
+	}
+
+	if kb.Spec.HTTP.TLS.Enabled() {
+		httpCertsVolume := certificates.HTTPCertSecretVolume(Namer, kb.Name)
+		volumes = append(volumes, httpCertsVolume)
+	}
+	return volumes
 }
 
 func NewService(kb kbv1.Kibana) *corev1.Service {

--- a/pkg/controller/kibana/driver_test.go
+++ b/pkg/controller/kibana/driver_test.go
@@ -246,9 +246,9 @@ func TestDriverDeploymentParams(t *testing.T) {
 			},
 			want: func() deployment.Params {
 				params := expectedDeploymentParams()
-				params.PodTemplateSpec.Spec.Volumes = append(params.PodTemplateSpec.Spec.Volumes[:1], params.PodTemplateSpec.Spec.Volumes[2:]...)
-				params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts = append(params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts[:1], params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts[2:]...)
-				params.PodTemplateSpec.Spec.Containers[0].VolumeMounts = append(params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[:1], params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[2:]...)
+				params.PodTemplateSpec.Spec.Volumes = params.PodTemplateSpec.Spec.Volumes[1:]
+				params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts = params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts[1:]
+				params.PodTemplateSpec.Spec.Containers[0].VolumeMounts = params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[1:]
 				params.PodTemplateSpec.Spec.Containers[0].ReadinessProbe.Handler.HTTPGet.Scheme = corev1.URISchemeHTTP
 				params.PodTemplateSpec.Spec.Containers[0].Ports[0].Name = "http"
 				return params
@@ -448,19 +448,19 @@ func expectedDeploymentParams() deployment.Params {
 			Spec: corev1.PodSpec{
 				Volumes: []corev1.Volume{
 					{
-						Name: "elastic-internal-elasticsearch-config",
+						Name: certificates.HTTPCertificatesSecretVolumeName,
 						VolumeSource: corev1.VolumeSource{
 							Secret: &corev1.SecretVolumeSource{
-								SecretName: "test-kb-config",
+								SecretName: "test-kb-http-certs-internal",
 								Optional:   &false,
 							},
 						},
 					},
 					{
-						Name: certificates.HTTPCertificatesSecretVolumeName,
+						Name: "elastic-internal-kibana-config",
 						VolumeSource: corev1.VolumeSource{
 							Secret: &corev1.SecretVolumeSource{
-								SecretName: "test-kb-http-certs-internal",
+								SecretName: "test-kb-config",
 								Optional:   &false,
 							},
 						},
@@ -508,14 +508,14 @@ func expectedDeploymentParams() deployment.Params {
 					},
 					VolumeMounts: []corev1.VolumeMount{
 						{
-							Name:      "elastic-internal-elasticsearch-config",
-							ReadOnly:  true,
-							MountPath: InternalConfigVolumeMountPath,
-						},
-						{
 							Name:      certificates.HTTPCertificatesSecretVolumeName,
 							ReadOnly:  true,
 							MountPath: certificates.HTTPCertificatesSecretVolumeMountPath,
+						},
+						{
+							Name:      "elastic-internal-kibana-config",
+							ReadOnly:  true,
+							MountPath: InternalConfigVolumeMountPath,
 						},
 						ConfigSharedVolume.InitContainerVolumeMount(),
 						{
@@ -544,14 +544,14 @@ func expectedDeploymentParams() deployment.Params {
 				Containers: []corev1.Container{{
 					VolumeMounts: []corev1.VolumeMount{
 						{
-							Name:      "elastic-internal-elasticsearch-config",
-							ReadOnly:  true,
-							MountPath: InternalConfigVolumeMountPath,
-						},
-						{
 							Name:      certificates.HTTPCertificatesSecretVolumeName,
 							ReadOnly:  true,
 							MountPath: certificates.HTTPCertificatesSecretVolumeMountPath,
+						},
+						{
+							Name:      "elastic-internal-kibana-config",
+							ReadOnly:  true,
+							MountPath: InternalConfigVolumeMountPath,
 						},
 						ConfigSharedVolume.VolumeMount(),
 						{

--- a/pkg/controller/kibana/driver_test.go
+++ b/pkg/controller/kibana/driver_test.go
@@ -27,6 +27,7 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/certificates"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/deployment"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/settings"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/compare"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
@@ -245,8 +246,9 @@ func TestDriverDeploymentParams(t *testing.T) {
 			},
 			want: func() deployment.Params {
 				params := expectedDeploymentParams()
-				params.PodTemplateSpec.Spec.Volumes = params.PodTemplateSpec.Spec.Volumes[:3]
-				params.PodTemplateSpec.Spec.Containers[0].VolumeMounts = params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[:3]
+				params.PodTemplateSpec.Spec.Volumes = append(params.PodTemplateSpec.Spec.Volumes[:1], params.PodTemplateSpec.Spec.Volumes[2:]...)
+				params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts = append(params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts[:1], params.PodTemplateSpec.Spec.InitContainers[0].VolumeMounts[2:]...)
+				params.PodTemplateSpec.Spec.Containers[0].VolumeMounts = append(params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[:1], params.PodTemplateSpec.Spec.Containers[0].VolumeMounts[2:]...)
 				params.PodTemplateSpec.Spec.Containers[0].ReadinessProbe.Handler.HTTPGet.Scheme = corev1.URISchemeHTTP
 				params.PodTemplateSpec.Spec.Containers[0].Ports[0].Name = "http"
 				return params
@@ -446,25 +448,10 @@ func expectedDeploymentParams() deployment.Params {
 			Spec: corev1.PodSpec{
 				Volumes: []corev1.Volume{
 					{
-						Name: DataVolumeName,
-						VolumeSource: corev1.VolumeSource{
-							EmptyDir: &corev1.EmptyDirVolumeSource{},
-						},
-					},
-					{
-						Name: "config",
+						Name: "elastic-internal-elasticsearch-config",
 						VolumeSource: corev1.VolumeSource{
 							Secret: &corev1.SecretVolumeSource{
 								SecretName: "test-kb-config",
-								Optional:   &false,
-							},
-						},
-					},
-					{
-						Name: "elasticsearch-certs",
-						VolumeSource: corev1.VolumeSource{
-							Secret: &corev1.SecretVolumeSource{
-								SecretName: "es-ca-secret",
 								Optional:   &false,
 							},
 						},
@@ -478,28 +465,104 @@ func expectedDeploymentParams() deployment.Params {
 							},
 						},
 					},
+					{
+						Name: ConfigSharedVolume.VolumeName,
+						VolumeSource: corev1.VolumeSource{
+							EmptyDir: &corev1.EmptyDirVolumeSource{},
+						},
+					},
+					{
+						Name: "elasticsearch-certs",
+						VolumeSource: corev1.VolumeSource{
+							Secret: &corev1.SecretVolumeSource{
+								SecretName: "es-ca-secret",
+								Optional:   &false,
+							},
+						},
+					},
+					{
+						Name: DataVolumeName,
+						VolumeSource: corev1.VolumeSource{
+							EmptyDir: &corev1.EmptyDirVolumeSource{},
+						},
+					},
 				},
-				Containers: []corev1.Container{{
+				InitContainers: []corev1.Container{{
+					Name:            "elastic-internal-init-config",
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					Image:           "my-image",
+					Command:         []string{"/usr/bin/env", "bash", "-c", InitConfigScript},
+					SecurityContext: &corev1.SecurityContext{
+						Privileged: &false,
+					},
+					Env: []corev1.EnvVar{
+						{Name: settings.EnvPodIP, Value: "", ValueFrom: &corev1.EnvVarSource{
+							FieldRef: &corev1.ObjectFieldSelector{APIVersion: "v1", FieldPath: "status.podIP"},
+						}},
+						{Name: settings.EnvPodName, Value: "", ValueFrom: &corev1.EnvVarSource{
+							FieldRef: &corev1.ObjectFieldSelector{APIVersion: "v1", FieldPath: "metadata.name"},
+						}},
+						{Name: settings.EnvNodeName, Value: "", ValueFrom: &corev1.EnvVarSource{
+							FieldRef: &corev1.ObjectFieldSelector{APIVersion: "v1", FieldPath: "spec.nodeName"},
+						}},
+					},
 					VolumeMounts: []corev1.VolumeMount{
 						{
-							Name:      DataVolumeName,
-							ReadOnly:  false,
-							MountPath: DataVolumeMountPath,
+							Name:      "elastic-internal-elasticsearch-config",
+							ReadOnly:  true,
+							MountPath: InternalConfigVolumeMountPath,
 						},
 						{
-							Name:      "config",
+							Name:      certificates.HTTPCertificatesSecretVolumeName,
 							ReadOnly:  true,
-							MountPath: "/usr/share/kibana/config",
+							MountPath: certificates.HTTPCertificatesSecretVolumeMountPath,
 						},
+						ConfigSharedVolume.InitContainerVolumeMount(),
 						{
 							Name:      "elasticsearch-certs",
 							ReadOnly:  true,
 							MountPath: "/usr/share/kibana/config/elasticsearch-certs",
 						},
 						{
+							Name:      DataVolumeName,
+							ReadOnly:  false,
+							MountPath: DataVolumeMountPath,
+						},
+					},
+					Resources: corev1.ResourceRequirements{
+						Requests: map[corev1.ResourceName]resource.Quantity{
+							corev1.ResourceMemory: resource.MustParse("50Mi"),
+							corev1.ResourceCPU:    resource.MustParse("0.1"),
+						},
+						Limits: map[corev1.ResourceName]resource.Quantity{
+							// Memory limit should be at least 12582912 when running with CRI-O
+							corev1.ResourceMemory: resource.MustParse("50Mi"),
+							corev1.ResourceCPU:    resource.MustParse("0.1"),
+						},
+					},
+				}},
+				Containers: []corev1.Container{{
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "elastic-internal-elasticsearch-config",
+							ReadOnly:  true,
+							MountPath: InternalConfigVolumeMountPath,
+						},
+						{
 							Name:      certificates.HTTPCertificatesSecretVolumeName,
 							ReadOnly:  true,
 							MountPath: certificates.HTTPCertificatesSecretVolumeMountPath,
+						},
+						ConfigSharedVolume.VolumeMount(),
+						{
+							Name:      "elasticsearch-certs",
+							ReadOnly:  true,
+							MountPath: "/usr/share/kibana/config/elasticsearch-certs",
+						},
+						{
+							Name:      DataVolumeName,
+							ReadOnly:  false,
+							MountPath: DataVolumeMountPath,
 						},
 					},
 					Image: "my-image",

--- a/pkg/controller/kibana/init_configuration.go
+++ b/pkg/controller/kibana/init_configuration.go
@@ -26,7 +26,7 @@ fi
 
 echo "Setup Kibana configuration"
 
-ln -sf /mnt/elastic-internal/kibana-config/* /mnt/elastic-internal/kibana-config-local/
+ln -sf ` + InternalConfigVolumeMountPath + `/* ` + InitContainerConfigVolumeMountPath + `/
 
 touch "${init_config_initialized_flag}"
 echo "Kibana configuration successfully prepared."

--- a/pkg/controller/kibana/init_configuration.go
+++ b/pkg/controller/kibana/init_configuration.go
@@ -33,8 +33,9 @@ echo "Kibana configuration successfully prepared."
 `
 )
 
-// initConfigContainer returns an init container that executes a bash script
-// to load secure settings in a Keystore.
+// initConfigContainer returns an init container that executes a bash script to prepare the Kibana config directory.
+// The script creates symbolic links from the generated configuration files in /mnt/elastic-internal/kibana-config/ to
+// an empty directory later mounted in /use/share/kibana/config
 func initConfigContainer(kb kbv1.Kibana) corev1.Container {
 	privileged := false
 

--- a/pkg/controller/kibana/init_configuration.go
+++ b/pkg/controller/kibana/init_configuration.go
@@ -1,0 +1,65 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kibana
+
+import (
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const (
+	InitConfigContainerName = "elastic-internal-init-config"
+
+	// InitConfigScript is a small bash script to prepare the Kibana configuration directory
+	InitConfigScript = `#!/usr/bin/env bash
+set -eux
+
+init_config_initialized_flag=/usr/share/kibana/config/elastic-internal-init-config.ok
+
+if [[ -f "${init_config_initialized_flag}" ]]; then
+    echo "Kibana configuration already initialized."
+	exit 0
+fi
+
+echo "Setup Kibana configuration"
+
+ln -sf /mnt/elastic-internal/kibana-config/* /mnt/elastic-internal/kibana-config-local/
+
+touch "${init_config_initialized_flag}"
+echo "Kibana configuration successfully prepared."
+`
+)
+
+// initConfigContainer returns an init container that executes a bash script
+// to load secure settings in a Keystore.
+func initConfigContainer(kb kbv1.Kibana) corev1.Container {
+	privileged := false
+
+	return corev1.Container{
+		// Image will be inherited from pod template defaults
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Name:            InitConfigContainerName,
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: &privileged,
+		},
+		Command: []string{"/usr/bin/env", "bash", "-c", InitConfigScript},
+		VolumeMounts: []corev1.VolumeMount{
+			ConfigSharedVolume.InitContainerVolumeMount(),
+			ConfigVolume(kb).VolumeMount(),
+		},
+		Resources: corev1.ResourceRequirements{
+			Requests: map[corev1.ResourceName]resource.Quantity{
+				corev1.ResourceMemory: resource.MustParse("50Mi"),
+				corev1.ResourceCPU:    resource.MustParse("0.1"),
+			},
+			Limits: map[corev1.ResourceName]resource.Quantity{
+				// Memory limit should be at least 12582912 when running with CRI-O
+				corev1.ResourceMemory: resource.MustParse("50Mi"),
+				corev1.ResourceCPU:    resource.MustParse("0.1"),
+			},
+		},
+	}
+}

--- a/pkg/controller/kibana/keystore.go
+++ b/pkg/controller/kibana/keystore.go
@@ -1,3 +1,7 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
 package kibana
 
 import (

--- a/pkg/controller/kibana/keystore.go
+++ b/pkg/controller/kibana/keystore.go
@@ -1,0 +1,43 @@
+package kibana
+
+import (
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/keystore"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// keystoreInConfigDirVersion is the version in which the keystore is no longer stored in the data directory but in the configuration one.
+var keystoreInConfigDirVersion = version.From(7, 9, 0)
+
+// newInitContainersParameters is used to generate the init container that will load the secure settings into a keystore
+func newInitContainersParameters(kb *kbv1.Kibana) (keystore.InitContainerParameters, error) {
+	parameters := keystore.InitContainerParameters{
+		KeystoreCreateCommand:         "/usr/share/kibana/bin/kibana-keystore create",
+		KeystoreAddCommand:            `/usr/share/kibana/bin/kibana-keystore add "$key" --stdin < "$filename"`,
+		SecureSettingsVolumeMountPath: keystore.SecureSettingsVolumeMountPath,
+		KeystoreVolumePath:            DataVolumeMountPath,
+		Resources: corev1.ResourceRequirements{
+			Requests: map[corev1.ResourceName]resource.Quantity{
+				corev1.ResourceMemory: resource.MustParse("128Mi"),
+				corev1.ResourceCPU:    resource.MustParse("100m"),
+			},
+			Limits: map[corev1.ResourceName]resource.Quantity{
+				corev1.ResourceMemory: resource.MustParse("128Mi"),
+				corev1.ResourceCPU:    resource.MustParse("100m"),
+			},
+		},
+	}
+
+	kbVersion, err := version.Parse(kb.Spec.Version)
+	if err != nil {
+		return parameters, err
+	}
+
+	if kbVersion.IsSameOrAfter(keystoreInConfigDirVersion) {
+		parameters.KeystoreVolumePath = ConfigSharedVolume.ContainerMountPath
+	}
+
+	return parameters, nil
+}

--- a/pkg/controller/kibana/keystore.go
+++ b/pkg/controller/kibana/keystore.go
@@ -12,7 +12,7 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 )
 
-// keystoreInConfigDirVersion is the version in which the keystore is no longer stored in the data directory but in the configuration one.
+// keystoreInConfigDirVersion is the version in which the keystore is no longer stored in the data directory but in the config one.
 var keystoreInConfigDirVersion = version.From(7, 9, 0)
 
 // newInitContainersParameters is used to generate the init container that will load the secure settings into a keystore

--- a/pkg/controller/kibana/pod.go
+++ b/pkg/controller/kibana/pod.go
@@ -16,7 +16,6 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/keystore"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/pod"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
-	commonvolume "github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
 )
 
 const (
@@ -70,7 +69,7 @@ func readinessProbe(useTLS bool) corev1.Probe {
 	}
 }
 
-func NewPodTemplateSpec(kb kbv1.Kibana, keystore *keystore.Resources, volumes []commonvolume.VolumeLike) corev1.PodTemplateSpec {
+func NewPodTemplateSpec(kb kbv1.Kibana, keystore *keystore.Resources, volumes []volume.VolumeLike) corev1.PodTemplateSpec {
 	labels := NewLabels(kb.Name)
 	labels[KibanaVersionLabelName] = kb.Spec.Version
 

--- a/pkg/controller/kibana/pod.go
+++ b/pkg/controller/kibana/pod.go
@@ -16,6 +16,7 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/keystore"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/pod"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
+	commonvolume "github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
 )
 
 const (
@@ -69,20 +70,11 @@ func readinessProbe(useTLS bool) corev1.Probe {
 	}
 }
 
-func NewPodTemplateSpec(kb kbv1.Kibana, keystore *keystore.Resources) corev1.PodTemplateSpec {
+func NewPodTemplateSpec(kb kbv1.Kibana, keystore *keystore.Resources, volumes []commonvolume.VolumeLike) corev1.PodTemplateSpec {
 	labels := NewLabels(kb.Name)
 	labels[KibanaVersionLabelName] = kb.Spec.Version
 
 	ports := getDefaultContainerPorts(kb)
-
-	volumes := []corev1.Volume{DataVolume.Volume()}
-	volumeMounts := []corev1.VolumeMount{DataVolume.VolumeMount()}
-	var initContainers []corev1.Container
-
-	if keystore != nil {
-		volumes = append(volumes, keystore.Volume)
-		initContainers = append(initContainers, keystore.InitContainer)
-	}
 
 	builder := defaults.NewPodTemplateBuilder(kb.Spec.PodTemplate, kbv1.KibanaContainerName).
 		WithResources(DefaultResources).
@@ -91,12 +83,18 @@ func NewPodTemplateSpec(kb kbv1.Kibana, keystore *keystore.Resources) corev1.Pod
 		WithDockerImage(kb.Spec.Image, container.ImageRepository(container.KibanaImage, kb.Spec.Version)).
 		WithReadinessProbe(readinessProbe(kb.Spec.HTTP.TLS.Enabled())).
 		WithPorts(ports).
-		WithVolumes(volumes...).
-		WithVolumeMounts(volumeMounts...).
-		WithInitContainers(initContainers...).
-		WithInitContainerDefaults()
+		WithInitContainers(initConfigContainer(kb))
 
-	return builder.PodTemplate
+	for _, volume := range volumes {
+		builder.WithVolumes(volume.Volume()).WithVolumeMounts(volume.VolumeMount())
+	}
+
+	if keystore != nil {
+		builder.WithVolumes(keystore.Volume).
+			WithInitContainers(keystore.InitContainer)
+	}
+
+	return builder.WithInitContainerDefaults().PodTemplate
 }
 
 // GetKibanaContainer returns the Kibana container from the given podSpec.

--- a/pkg/controller/kibana/pod_test.go
+++ b/pkg/controller/kibana/pod_test.go
@@ -7,6 +7,8 @@ package kibana
 import (
 	"testing"
 
+	commonvolume "github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
+
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
@@ -37,11 +39,11 @@ func TestNewPodTemplateSpec(t *testing.T) {
 			assertions: func(pod corev1.PodTemplateSpec) {
 				assert.Equal(t, false, *pod.Spec.AutomountServiceAccountToken)
 				assert.Len(t, pod.Spec.Containers, 1)
-				assert.Len(t, pod.Spec.InitContainers, 0)
-				assert.Len(t, pod.Spec.Volumes, 1)
+				assert.Len(t, pod.Spec.InitContainers, 1)
+				assert.Len(t, pod.Spec.Volumes, 0)
 				kibanaContainer := GetKibanaContainer(pod.Spec)
 				require.NotNil(t, kibanaContainer)
-				assert.Equal(t, 1, len(kibanaContainer.VolumeMounts))
+				assert.Equal(t, 0, len(kibanaContainer.VolumeMounts))
 				assert.Equal(t, container.ImageRepository(container.KibanaImage, "7.1.0"), kibanaContainer.Image)
 				assert.NotNil(t, kibanaContainer.ReadinessProbe)
 				assert.NotEmpty(t, kibanaContainer.Ports)
@@ -59,8 +61,8 @@ func TestNewPodTemplateSpec(t *testing.T) {
 				Volume:        corev1.Volume{Name: "vol"},
 			},
 			assertions: func(pod corev1.PodTemplateSpec) {
-				assert.Len(t, pod.Spec.InitContainers, 1)
-				assert.Len(t, pod.Spec.Volumes, 2)
+				assert.Len(t, pod.Spec.InitContainers, 2)
+				assert.Len(t, pod.Spec.Volumes, 1)
 			},
 		},
 		{
@@ -127,8 +129,9 @@ func TestNewPodTemplateSpec(t *testing.T) {
 			}},
 			keystore: nil,
 			assertions: func(pod corev1.PodTemplateSpec) {
-				assert.Len(t, pod.Spec.InitContainers, 1)
+				assert.Len(t, pod.Spec.InitContainers, 2)
 				assert.Equal(t, pod.Spec.Containers[0].Image, pod.Spec.InitContainers[0].Image)
+				assert.Equal(t, pod.Spec.Containers[0].Image, pod.Spec.InitContainers[1].Image)
 			},
 		},
 		{
@@ -206,14 +209,16 @@ func TestNewPodTemplateSpec(t *testing.T) {
 				},
 			}},
 			assertions: func(pod corev1.PodTemplateSpec) {
-				assert.Len(t, pod.Spec.Volumes, 2)
-				assert.Len(t, GetKibanaContainer(pod.Spec).VolumeMounts, 2)
+				assert.Len(t, pod.Spec.InitContainers, 1)
+				assert.Len(t, pod.Spec.InitContainers[0].VolumeMounts, 3)
+				assert.Len(t, pod.Spec.Volumes, 1)
+				assert.Len(t, GetKibanaContainer(pod.Spec).VolumeMounts, 1)
 			},
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got := NewPodTemplateSpec(tt.kb, tt.keystore)
+			got := NewPodTemplateSpec(tt.kb, tt.keystore, []commonvolume.VolumeLike{})
 			tt.assertions(got)
 		})
 	}

